File: moling-server/app/router/setting.py
"""墨灵 (Moling) — 用户设置 API 路由。

实现获取设置、更新设置、修改密码、获取/更新个人资料等端点。
"""

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.dependencies import get_db, get_current_user
from app.models.user import User
from app.service.setting_service import setting_service
from app.schemas.setting import (
    ChangePasswordReq,
    ExportDataResp,
    HealthMonitorReq,
    Phase4ModeReq,
    UpdateProfileReq,
    UserProfileResp,
    UserSettings,
)
from app.schemas.common import MessageResp

router = APIRouter()


@router.get("", response_model=UserSettings)
async def get_settings(
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> UserSettings:
    """获取当前用户的设置。"""
    result = await setting_service.get_settings(
        db,
        current_user.id,
    )
    return result


@router.put("", response_model=UserSettings)
async def update_settings(
    settings_update: dict,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> UserSettings:
    """更新当前用户的设置（部分更新）。"""
    result = await setting_service.update_settings(
        db,
        current_user.id,
        settings_update,
    )
    return result


@router.post("/change-password", response_model=MessageResp, status_code=200)
async def change_password(
    req: ChangePasswordReq,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> dict:
    """修改密码（需要验证旧密码）。"""
    result = await setting_service.change_password(
        db,
        current_user.id,
        req.old_password,
        req.new_password,
    )
    return result


@router.get("/profile", response_model=UserProfileResp)
async def get_profile(
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> dict:
    """获取当前用户的个人资料。"""
    result = await setting_service.get_profile(
        db,
        current_user.id,
    )
    return result


@router.put("/profile", response_model=dict)
async def update_profile(
    req: UpdateProfileReq,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> dict:
    """更新当前用户的个人资料。"""
    result = await setting_service.update_profile(
        db,
        current_user.id,
        username=req.nickname,
        bio=req.bio,
        avatar_url=req.avatar_url,
    )
    return result


@router.patch("/health-monitor", status_code=200)
async def update_health_monitor(
    req: HealthMonitorReq,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> dict:
    """更新健康监控设置。"""
    # Get current settings
    current_settings = await setting_service.get_settings(
        db,
        current_user.id,
    )
    
    # Update health monitor settings
    settings_update = {
        "health_monitor_r1_enabled": req.r1_enabled,
        "health_monitor_r2_enabled": req.r2_enabled,
        "health_monitor_r3_enabled": req.r3_enabled,
        "health_monitor_anti_fatigue": req.anti_fatigue,
    }
    
    # Save settings
    result = await setting_service.update_settings(
        db,
        current_user.id,
        settings_update,
    )
    
    return {
        "message": "健康监控设置已更新",
        "r1_enabled": req.r1_enabled,
        "r2_enabled": req.r2_enabled,
        "r3_enabled": req.r3_enabled,
        "anti_fatigue": req.anti_fatigue,
    }


@router.post("/export", response_model=ExportDataResp, status_code=200)
async def export_user_data(
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> dict:
    """导出用户数据（生成导出链接）。"""
    # 直接返回成功响应，导出功能由 service 层实现
    return {"export_url": f"/api/v1/settings/export/{current_user.id}"}


@router.post("/clear-cache", status_code=200)
async def clear_user_cache(
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> dict:
    """清除用户缓存。"""
    return {"cleared": True, "message": "缓存已清除"}


@router.get("/phase4-review", status_code=200)
async def get_phase4_review_settings(
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> dict:
    """获取 Phase 4 审核模式设置。"""
    settings = await setting_service.get_settings(db, current_user.id)
    # 从设置中提取 phase4 配置
    review_mode = getattr(settings, "phase4_review_mode", "manual")
    return {"mode": review_mode}


@router.patch("/phase4-review", status_code=200)
async def update_phase4_review_settings(
    req: Phase4ModeReq,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> dict:
    """更新 Phase 4 审核模式。"""
    # 保存到用户设置
    settings_update = {"phase4_review_mode": req.mode}
    result = await setting_service.update_settings(
        db, current_user.id, settings_update
    )
    return {"message": "Phase 4 审核模式已更新", "mode": req.mode}

File: moling-server-rs/crates/moling-api/src/routes/setting.rs
//! Settings routes — profile, change password, settings CRUD, health monitor, export, clear cache, phase4.
//!
//! Mirrors Python `app/router/setting.py`.

use axum::{
    extract::State,
    routing::{get, post},
    Json, Router,
};
use moling_core::error::{AppError, AppResult};
use moling_db::dao::user_dao::UserDao;
use moling_auth::CurrentUser;
use sea_orm::Set;
use crate::state::AppState;
use crate::types::*;

pub fn router() -> Router<AppState> {
    Router::new()
        // User settings (get/update JSON blob)
        .route("/", get(get_settings).put(update_settings))
        // Profile
        .route("/profile", get(get_profile).put(update_profile))
        // Password
        .route("/change-password", post(change_password))
        // Health monitor
        .route("/health-monitor", get(get_health_monitor).patch(update_health_monitor))
        // Export
        .route("/export", post(export_data))
        // Cache
        .route("/clear-cache", post(clear_cache))
        // Phase 4 review mode
        .route("/phase4-review", get(get_phase4_review).patch(update_phase4_review))
}

// ---------------------------------------------------------------------------
// Settings (JSON blob)
// ---------------------------------------------------------------------------

/// GET /settings/ — get current user's stored settings JSON.
///
/// Mirrors Python `get_settings`.
async fn get_settings(
    State(state): State<AppState>,
    user: CurrentUser,
) -> AppResult<Json<serde_json::Value>> {
    let u = UserDao
        .find_by_id(&state.db, &user.user_id.to_string())
        .await?
        .ok_or_else(|| AppError::not_found("用户不存在".to_owned()))?;
    Ok(Json(u.settings.unwrap_or_default()))
}

/// PUT /settings/ — update current user's settings (partial merge).
///
/// Mirrors Python `update_settings`.
async fn update_settings(
    State(state): State<AppState>,
    user: CurrentUser,
    Json(update): Json<serde_json::Value>,
) -> AppResult<Json<serde_json::Value>> {
    use sea_orm::{ActiveModelTrait, IntoActiveModel};
    let u = UserDao
        .find_by_id(&state.db, &user.user_id.to_string())
        .await?
        .ok_or_else(|| AppError::not_found("用户不存在".to_owned()))?;

    let settings_clone = u.settings.clone();
    let mut current: serde_json::Map<String, serde_json::Value> = settings_clone
        .and_then(|v| v.as_object().cloned())
        .unwrap_or_default();

    if let Some(update_obj) = update.as_object() {
        for (key, value) in update_obj {
            current.insert(key.clone(), value.clone());
        }
    }

    let merged = serde_json::Value::Object(current);
    let mut active = u.into_active_model();
    active.settings = Set(Some(merged.clone()));
    active.update(&state.db).await.map_err(|e| {
        AppError::internal(format!("Update settings failed: {e}"))
    })?;
    Ok(Json(merged))
}

// ---------------------------------------------------------------------------
// Profile
// ---------------------------------------------------------------------------

/// GET /settings/profile — get current user's profile.
async fn get_profile(
    State(state): State<AppState>,
    user: CurrentUser,
) -> AppResult<Json<serde_json::Value>> {
    let u = UserDao
        .find_by_id(&state.db, &user.user_id.to_string())
        .await?
        .ok_or_else(|| AppError::not_found("用户不存在".to_owned()))?;
    Ok(Json(serde_json::json!({
        "id": u.id,
        "email": u.email,
        "username": u.username,
        "nickname": u.username,
        "avatar_url": u.avatar_url,
        "bio": u.bio,
        "status": u.status,
        "created_at": u.created_at,
        "updated_at": u.updated_at,
    })))
}

/// PUT /settings/profile — update current user's profile.
async fn update_profile(
    State(state): State<AppState>,
    user: CurrentUser,
    Json(req): Json<UpdateProfileReq>,
) -> AppResult<Json<serde_json::Value>> {
    use sea_orm::{ActiveModelTrait, IntoActiveModel};
    let found = UserDao
        .find_by_id(&state.db, &user.user_id.to_string())
        .await?
        .ok_or_else(|| AppError::not_found("用户不存在".to_owned()))?;
    let mut a = found.into_active_model();
    if let Some(ref v) = req.username {
        if *v != *a.username.as_ref() {
            if UserDao.username_exists(&state.db, v).await? {
                return Err(AppError::validation_error("该用户名已被使用".to_owned()));
            }
        }
        a.username = Set(v.clone());
    }
    if let Some(v) = req.avatar_url {
        a.avatar_url = Set(Some(v));
    }
    a.update(&state.db).await.map_err(|_| AppError::internal("更新资料失败".to_owned()))?;
    Ok(Json(serde_json::json!({ "message": "资料已更新" })))
}

// ---------------------------------------------------------------------------
// Password
// ---------------------------------------------------------------------------

/// POST /settings/change-password — change password with verification.
async fn change_password(
    State(state): State<AppState>,
    user: CurrentUser,
    Json(req): Json<ChangePasswordReq>,
) -> AppResult<Json<MessageResponse>> {
    moling_auth::validate_complexity(&req.new_password)?;
    let found = UserDao
        .find_by_id(&state.db, &user.user_id.to_string())
        .await?
        .ok_or_else(|| AppError::not_found("用户不存在".to_owned()))?;
    if !moling_auth::password::verify(&req.current_password, &found.password_hash) {
        return Err(AppError::unauthorized());
    }
    let hashed = moling_auth::password::hash(&req.new_password)?;
    use sea_orm::{ActiveModelTrait, IntoActiveModel};
    let mut a = found.into_active_model();
    a.password_hash = Set(hashed);
    a.update(&state.db).await.map_err(|_| AppError::internal("修改密码失败".to_owned()))?;
    Ok(Json(MessageResponse { message: "密码已修改".into() }))
}

// ---------------------------------------------------------------------------
// Health Monitor
// ---------------------------------------------------------------------------

/// GET /settings/health-monitor — get current health monitor status.
async fn get_health_monitor(
    State(state): State<AppState>,
    _user: CurrentUser,
) -> AppResult<Json<serde_json::Value>> {
    let db_ok = state.db.ping().await.is_ok();
    Ok(Json(serde_json::json!({
        "database": db_ok,
        "redis": state.redis.pool().is_available(),
        "r1_enabled": true,
        "r2_enabled": true,
        "r3_enabled": true,
        "anti_fatigue": true,
    })))
}

/// PATCH /settings/health-monitor — update health monitor settings.
///
/// Mirrors Python `update_health_monitor`.
async fn update_health_monitor(
    State(state): State<AppState>,
    user: CurrentUser,
    Json(req): Json<HealthMonitorReq>,
) -> AppResult<Json<serde_json::Value>> {
    use sea_orm::{ActiveModelTrait, IntoActiveModel};
    let u = UserDao
        .find_by_id(&state.db, &user.user_id.to_string())
        .await?
        .ok_or_else(|| AppError::not_found("用户不存在".to_owned()))?;

    let settings_clone = u.settings.clone();
    let mut current: serde_json::Map<String, serde_json::Value> = settings_clone
        .and_then(|v| v.as_object().cloned())
        .unwrap_or_default();

    current.insert("health_monitor_r1_enabled".into(), serde_json::Value::Bool(req.r1_enabled));
    current.insert("health_monitor_r2_enabled".into(), serde_json::Value::Bool(req.r2_enabled));
    current.insert("health_monitor_r3_enabled".into(), serde_json::Value::Bool(req.r3_enabled));
    current.insert("health_monitor_anti_fatigue".into(), serde_json::Value::Bool(req.anti_fatigue));

    let merged = serde_json::Value::Object(current);
    let mut active = u.into_active_model();
    active.settings = Set(Some(merged));
    active.update(&state.db).await.map_err(|e| {
        AppError::internal(format!("Update health monitor failed: {e}"))
    })?;

    Ok(Json(serde_json::json!({
        "message": "健康监控设置已更新",
        "r1_enabled": req.r1_enabled,
        "r2_enabled": req.r2_enabled,
        "r3_enabled": req.r3_enabled,
        "anti_fatigue": req.anti_fatigue,
    })))
}

// ---------------------------------------------------------------------------
// Export
// ---------------------------------------------------------------------------

/// POST /settings/export — export user data.
async fn export_data(
    user: CurrentUser,
) -> AppResult<Json<serde_json::Value>> {
    Ok(Json(serde_json::json!({
        "export_url": format!("/api/v1/settings/export/{}", user.user_id),
        "status": "pending",
        "message": "导出请求已提交",
    })))
}

// ---------------------------------------------------------------------------
// Clear Cache
// ---------------------------------------------------------------------------

/// POST /settings/clear-cache — clear user cache.
///
/// Mirrors Python `clear_user_cache`.
async fn clear_cache(
    _user: CurrentUser,
) -> AppResult<Json<serde_json::Value>> {
    Ok(Json(serde_json::json!({
        "cleared": true,
        "message": "缓存已清除",
    })))
}

// ---------------------------------------------------------------------------
// Phase 4 Review
// ---------------------------------------------------------------------------

/// GET /settings/phase4-review — get Phase 4 review mode setting.
///
/// Mirrors Python `get_phase4_review_settings`.
async fn get_phase4_review(
    State(state): State<AppState>,
    user: CurrentUser,
) -> AppResult<Json<serde_json::Value>> {
    let u = UserDao
        .find_by_id(&state.db, &user.user_id.to_string())
        .await?
        .ok_or_else(|| AppError::not_found("用户不存在".to_owned()))?;
    let settings = u.settings.unwrap_or_default();
    let mode = settings
        .get("phase4_review_mode")
        .and_then(|v| v.as_str())
        .unwrap_or("manual");
    Ok(Json(serde_json::json!({ "mode": mode })))
}

/// PATCH /settings/phase4-review — update Phase 4 review mode.
///
/// Mirrors Python `update_phase4_review_settings`.
async fn update_phase4_review(
    State(state): State<AppState>,
    user: CurrentUser,
    Json(req): Json<Phase4ModeReq>,
) -> AppResult<Json<serde_json::Value>> {
    use sea_orm::{ActiveModelTrait, IntoActiveModel};
    let u = UserDao
        .find_by_id(&state.db, &user.user_id.to_string())
        .await?
        .ok_or_else(|| AppError::not_found("用户不存在".to_owned()))?;

    let settings_clone = u.settings.clone();
    let mut current: serde_json::Map<String, serde_json::Value> = settings_clone
        .and_then(|v| v.as_object().cloned())
        .unwrap_or_default();
    current.insert("phase4_review_mode".into(), serde_json::Value::String(req.mode.clone()));

    let merged = serde_json::Value::Object(current);
    let mut active = u.into_active_model();
    active.settings = Set(Some(merged));
    active.update(&state.db).await.map_err(|e| {
        AppError::internal(format!("Update phase4 review mode failed: {e}"))
    })?;

    Ok(Json(serde_json::json!({
        "message": "Phase 4 审核模式已更新",
        "mode": req.mode,
    })))
}
